### PR TITLE
Fix/Newsfeed performance

### PR DESCRIPTION
**Pull request by roncoderpush**
## Overview
Work on this Trello card: https://trello.com/c/0j4LG3GV/196-android-newsfeed-lagging-chopping
Details:
- Fixing pull to refresh causing lagging
- Fixing scrooling too fast make scroll position back to 1
- Fixing newsfeed refreshed when come back from another screen
- Fixing when create a new post it will not be display

## How:
- [**Pull to refresh causing lagging and reset scroll position**]
-> I have implemented Smart Refresher which maintain scroll position, also changing bloc base on different states to one state but different status. This action keep all the UI that has been build remain the same and only add new UI to the screen, make the app no longer lagging.
- [**Fixing newsfeed refreshed when come back from another screen**] 
-> Only call refresh newsfeed when requested (pull to refresh, scroll to load more)
- [**Fixing when create a new post it will not be display**] 
-> Add PostDto we received after create new post to the top of the list.

## Screenshot
Simulator Screenshot - iPhone 14 Pro Max - 2023-09-06 at 14 20 07

Simulator Screenshot - iPhone 14 Pro Max - 2023-09-06 at 14 34 27

### CODE REVIEW

File: lib/core/application/post/create_post_bloc/create_post_bloc.dart [hunk 1]
```diff
@@ -79,8 +80,12 @@ class CreatePostBloc extends Cubit<CreatePostState> {
     );
     response.fold(
       (l) => emit(state.copyWith(status: CreatePostStatus.error)),
-      (isUpdateSuccess) =>
-          emit(state.copyWith(status: CreatePostStatus.postCreated)),
+      (postDto) => emit(
+        state.copyWith(
+          status: CreatePostStatus.postCreated,
+          newPost: Post.fromDto(postDto),
```
Comment (mthinh) [inline lib/core/application/post/create_post_bloc/create_post_bloc.dart:86]: the convert from dto to entities should be handled in repository not in bloc

File: lib/core/data/post/post_repository_impl.dart [hunk 2]
```diff
@@ -66,7 +66,7 @@ class PostRepositoryImpl implements PostRepository {
     if (result.hasException) {
       return Left(Failure());
     }
-    return Right(result.parsedData == true);
+    return Right(PostDto.fromJson(result.parsedData));
```
Comment (mthinh) [inline lib/core/data/post/post_repository_impl.dart:69]: Repository should only return entities not dto. please convert from dto to entites in parseFn of gql

File: lib/core/domain/post/post_repository.dart [hunk 3]
```diff
@@ -9,7 +10,7 @@ import '../../application/post/create_post_bloc/create_post_bloc.dart';
 abstract class PostRepository {
   Future<Either<Failure, List<Post>>> getPosts({GetPostsInput? input});
 
-  Future<Either<Failure, bool>> createNewPost({
+  Future<Either<Failure, PostDto>> createNewPost({
```
Comment (mthinh) [inline lib/core/domain/post/post_repository.dart:13]: The repository should only return entities

File: lib/core/service/post/post_service.dart [hunk 4]
```diff
@@ -16,7 +17,7 @@ class PostService {
     return postRepository.getPosts(input: input);
   }
 
-  Future<Either<Failure, bool>> createPost({
+  Future<Either<Failure, PostDto>> createPost({
```
Comment (mthinh) [inline lib/core/service/post/post_service.dart:20]: use entities

**Review by mthinh — CHANGES_REQUESTED** [review]
Please help check

File: lib/core/application/post/create_post_bloc/create_post_bloc.dart (hunk 1 above)
Comment (sieuhuflit) [reply lib/core/application/post/create_post_bloc/create_post_bloc.dart:86]: Yep, convert from dt to entited shouldn't handle here 👍

**Comment by roncoderpush** [comment]
@sieuhuflit @mthinh changes applied, kindly check it again

**Review by sieuhuflit — APPROVED** [review]

**Pull request merged**